sources/api/early-boot-config/src/provider/vmware.rs

```diff
@@ -0,0 +1,392 @@
+                // (this isn't perfect because chars aren't grapheme clusters, but will error
+                // toward printing the whole input, which is fine)
+                String::from_iter(user_data_str.chars().take(2048))
+            );
+        }
+
+        let json = SettingsJson::from_toml_str(&user_data_str, "user data")?;
+
+        Ok(Some(json))
+    }
+
+    /// Read and base64 decode user data contained in an OVF file
+    // In VMWare, user data is supplied to the host via an XML file.  Within
+    // the XML file, there is a `PropertySection` that contains `Property` elements
+    // with attributes.  User data is base64 encoded inside a `Property` element with
+    // the attribute "user-data".
+    // <Property key="user-data" value="1234abcd"/>
+    fn ovf_user_data<P: AsRef<Path>>(path: P) -> Result<String> {
+        let path = path.as_ref();
+        let file = File::open(path).context(error::InputFileRead { path })?;
+        let reader = OptionalCompressionReader::new(BufReader::new(file));
+
+        // Deserialize the OVF file, dropping everything we don't care about
+        let ovf: Environment =
+            serde_xml_rs::from_reader(reader).context(error::XmlDeserialize { path })?;
+
+        // We have seen the keys in the `Property` section be "namespaced" like "oe:key" or
+        // "of:key".  Since we aren't trying to validate the schema beyond the presence of the
+        // elements we care about, we can ignore the namespacing.  An example of this type of
+        // namespacing can be found in the unit test sample data. `serde_xml_rs` effectively
+        // ignores these namespaces and returns "key" / "value":
+        // https://github.com/Rreverser/serde-xml-rs/issues/64#issuecomment=540448434
+        let mut base64_str = String::new();
+        let user_data_key = "user-data";
+        for property in ovf.property_section.properties {
+            if property.key == user_data_key {
+                base64_str = property.value;
+                break;
+            }
+        }
+
+        // Base64 decode the &str
+        let decoded_bytes = base64::decode(&base64_str).context(error::Base64Decode {
+            base64_string: base64_str.to_string(),
+        })?;
+
+        // Decompress the data if it's compressed
+        let decoded = expand_slice_maybe(&decoded_bytes).context(error::Decompression {
+            what: "OVF user data",
+        })?;
+
+        Ok(decoded)
+    }
+
+    /// Read and decode user data based on values retrieved from the guestinfo interface
+    fn guestinfo_user_data() -> Result<Option<SettingsJson>> {
+        info!("Attempting to retrieve user data via guestinfo interface");
+
+        // Request a key's value from guestinfo
+        // TODO: Figure out if it's possible to re-use the backdoor or ERPC channel (mutable
+        // references everywhere...)
+        fn backdoor_get_bytes(key: &str) -> Result<Option<Vec<u8>>> {
```

Comment:
The function is nested because I would never expect (or want to) use it outside of the current function.  I think when some of these "data provider" methods are factored out a bit that will change, but until then it's safely located where it can't be misused. :)

Re: the TODO: My first iteration of this code created a backdoor and ERPC channel, then attempted to use it to pull the encoding and userdata.  The first call would succeed, the second call would fail, almost like the channel was "used up".  As soon as I instantiated a new backdoor/channel for the second call, everything was happy.  I've never used an ERPC channel like this before and didn't get too far into the source code to follow all the mutable references and see what's happening, so I left the TODO there as a reminder that it may not be necessary to create a backdoor/channel for each call.  There are only 2 calls, so I think it's ok for now.

Reply:
It is common to break out private functions even if they are only called in one place. It makes it easier to read. Nesting the function seems very unidiomatic to me. I very strongly advocate for moving it outside of its calling function.

For precedent, look at many of our main.rs files, where we have functions that are called only once! i.e. They exist purely for human readability and don't even facilitate code re-use.

The human readability factor works like this `backdoor_get_bytes` is named clearly enough for me to understand that I'm getting some bytes from somewhere, so I can follow the logic of `guestinfo_user_data` without concerning myself about where the bytes are coming from.  Unless, I do care, in which case I go look at `backdoor_get_bytes` and see a crisp isolated function that tells me where the bytes come from.